File: src/main/java/com/github/zxh/classpy/classfile/reader/BytesReader.java
package com.github.zxh.classpy.classfile.reader;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class BytesReader {
    
    protected final ByteBuffer buf;

    public BytesReader(byte[] bytes, ByteOrder order, boolean readOnly) {
        ByteBuffer _buf = ByteBuffer.wrap(bytes);
        this.buf = readOnly
                ? _buf.asReadOnlyBuffer().order(order)
                : _buf.order(order);
    }
    
    public ByteBuffer getByteBuffer() {
        return buf;
    }
    
    public int getPosition() {
        return buf.position();
    }
    
    public void skipBytes(int n) {
        for (int i = 0; i < n; i++) {
            buf.get();
        }
    }
    
    // byte[]
    public byte[] readBytes(int n) {
        byte[] bytes = new byte[n];
        buf.get(bytes);
        return bytes;
    }
    
    // 8-bit signed int
    public byte readByte() {
        return buf.get();
    }
    
    // 8-bit unsigned int
    public int readUnsignedByte() {
        return Byte.toUnsignedInt(buf.get());
    }
    
    // 16-bit signed int
    public short readShort() {
        return buf.getShort();
    }
    
    // 16-bit unsigned int
    public int readUnsignedShort() {
        return Short.toUnsignedInt(buf.getShort());
    }
    
    // 32-bit signed int
    public int readInt() {
        return buf.getInt();
    }

}

File: src/main/java/com/github/zxh/classpy/classfile/reader/BytesReader.kt
package com.github.zxh.classpy.classfile.reader

import java.nio.ByteBuffer
import java.nio.ByteOrder

open class BytesReader(bytes: ByteArray) {
    
    private val buf: ByteBuffer = ByteBuffer.wrap(bytes)
            .asReadOnlyBuffer()
            .order(ByteOrder.BIG_ENDIAN)

    val position: Int
        get() = buf.position()

    fun getByte(index: Int): Byte  = buf.get(index)
    fun getShort(index: Int): Short  = buf.getShort(index)
    
    fun readByte(): Byte = buf.get()
    fun readUnsignedByte(): Int = java.lang.Byte.toUnsignedInt(buf.get())
    fun readShort(): Short = buf.getShort()
    fun readUnsignedShort(): Int = java.lang.Short.toUnsignedInt(buf.getShort())
    fun readInt(): Int = buf.getInt()

    fun readBytes(n: Int): ByteArray {
        val bytes = ByteArray(n);
        buf.get(bytes);
        return bytes;
    }

    fun skipBytes(n: Int) {
        for (i in 0..n-1) {
            buf.get();
        }
    }

}
